### ISSUE TITLE
When a user replies to themselves, replies to *that* reply are (sometimes?) placed in the wrong location

### ISSUE THREAD

**Issue by afontenot** [mastodon/mastodon#21879]
### Steps to reproduce the problem

1. Check out a post by a user that contains a reply also by the same user. Preferably one with a lot of comments to the upper level post.
2. Look through the thread with the upper level post active and see if you can find the reply to the lower level post.

### Expected behaviour

Either the reply isn't shown at all, or it is shown at the normal location for replies.

### Actual behaviour

It is shown at a seemingly random location?

### Detailed description

Example of a thread that illustrates this issue (visible even when logged out): https://mastodon.social/@Meyerweb/109430741858138880

Walkthrough of the issue with screenshots:

<details><summary>The top level reply has a reply by the same user, and that reply has a reply</summary>

![example_part1](https://user-images.githubusercontent.com/3870410/204770209-05bd97fe-ebca-4b25-bd2e-c41e819f66dc.png)

</details>

<details><summary>The reply to the second level comment (in this case there's only one)</summary>

![example_part2](https://user-images.githubusercontent.com/3870410/204770366-b7dd9b86-0dfd-40da-a68a-fe6fd42a94ee.png)

</details>

Now, clicking back to the top level post, and looking through the replies, I found this reply in almost the exact middle. 

<details><summary>Where I found the reply</summary>

![example_part3](https://user-images.githubusercontent.com/3870410/204770688-2d3da0e2-7d9d-4e54-bc87-90ffde28ad78.png)

</details>

This placement doesn't seem to make any sense at all. It's entirely decontextualized -- I don't see how it's even possible to guess what post it's replying to. Seems like this has to be caused by a bug.

The normal placement for replies would dictate that it should go immediately after the post it's replying to.

### Specifications

Mastodon 4.0.2 (mastodon.social)
Firefox

**Comment by afontenot** [comment]
Interestingly, this might not be a front-end issue *per se*. Fetching the context JSON, I noticed that the post appears in the same position in the `descendants` item as it does in the web UI. Does Mastodon assume *ordered* JSON? That would be strange, to say the least.

**Comment by trwnh** [comment]
The flattening is done chronologically and not by subtrees. The "seemingly random" reply was a reply to the top-level post in the thread.

Worth noting that Pleroma returns the *entire* thread context (including all earlier posts, not just direct ancestors), and its UI is a bit more "imageboard"-y with reply links at the bottom of each post: 

![image](https://user-images.githubusercontent.com/10606431/204803550-df688ba1-3d3f-422f-94dc-35d2b2a1057d.png)

Akkoma UI has an option to show trees:

![image](https://user-images.githubusercontent.com/10606431/204803869-0448f5e7-ce15-4b7f-bc13-33126ea09f0b.png)

**Comment by afontenot** [comment]
> The flattening is done chronologically and not by subtrees. The "seemingly random" reply was a reply to the top-level post in the thread.

No, the reply I show in the screenshot was not a reply to the top-level post, it was a reply to a reply - specifically the reply made by the user who made the top level post, replying to their own post. That's the confusing thing. I expect top level *replies* to be chronological, that part makes sense. If I'm not mistaken, though, posts *are* given in thread order usually. @trwnh could you have another look and confirm whether or not the behavior matches what you expect?

I did a quick test like this:

```
top level post
  second level post 1
    third level post replying to post 1
  second level post 2
```

I made the third level post *last* out of these. Despite that, when refreshing the top level post, it appears *third*, in the "thread order" indicated above.

**Comment by afontenot** [comment]
If I'm not mistaken, this is the exact behavior I see:

1. Descendants are listed immediately after their parents.
2. Siblings are listed immediately after descendants.
3. Siblings are sorted in chronological order.
4. Exception 1: replies by the top-post author are *always* sorted first compared to siblings
5. Exception 2: when the top-post author has replies to their own top-level post, responses *by other users* to those *replies* are treated as if they were direct descendants of the top-level post, and listed in chronological order with other "siblings" at the top level.

5 is the bug I'm referring to. I don't understand why these replies have been moved to the top level instead of being presented in the normal thread ordering represented by 1-4.

**Comment by afontenot** [comment]
I tracked the cause of this issue down to https://github.com/mastodon/mastodon/pull/9320 - the relevant code is basically unchanged since then. See https://github.com/mastodon/mastodon/blob/main/app/models/concerns/status_threading_concern.rb#L91

The reason this was changed was because of a user report https://github.com/mastodon/mastodon/issues/6463 that basically amounts to the complaint that Mastodon doesn't support thread authoring in the web interface; you have to compose your thread one post at a time with the result that if someone replies before you finish the thread, strict chronological ordering would make your thread hard to read. (By "thread" here I just mean a series of post by the same author - usually needed because of the character limit.)

Instead of implementing thread authoring as the solution, a workaround was implemented instead, where a user's replies that are directly connected to the top post automatically get promoted over other siblings. Strikes me as a bad way to solve the problem of not being able to author threads, but that's neither here nor there.

I suspect, though, that completely breaking context in the way that this does was unintended and should be regarded as a bug. After all, the use case that drove this change was someone taking a few minutes to compose a reply to their own post, but the result is that posts hours or days later get promoted regardless of user intent.

Possible solutions:

1. Keep thread order intact by moving *all descendants* of self-replies, not just the self-replies.
2. Hide non-moved replies to moved replies, requiring a click to see them. (There's no decontextualization problem if the posts aren't visible.)
3. Revert the change so that self-replies aren't moved, and solve the *original* problem by implementing thread authoring instead. (I think there's a feature request for this but I couldn't find it.)
4. Implement (1), but also only move self-replies that are posted within ~10 minutes (exact time TBD) of the top-post.
5. If more robust thread presentation were implemented (e.g. indentation with toggleable sub-threads), replies to moved self-replies could be moved as in (1) — and minimized by default — allowing them to be toggled without having to change the thread context.

----------------------------

Quick illustration of why this is a problem:

```
True chronological ordering:

A (9h ago): My mother is sick again. :-(
  B (8h ago): @A I'm sorry :-(
  A (7h ago): At least I got that raise today.
    C (6h ago): @A Well deserved!

Mastodon's flattened ordering:

A (9h ago): My mother is sick again. :-(
A (7h ago): At least I got that raise today.
B (8h ago): @A I'm sorry :-(
C (6h ago): @A Well deserved!
```

Obviously this is a little contrived, but with a large enough or confusing thread this is just asking for misunderstandings to happen.
